src/game_objects/player.tbasicu

```diff
@@ -107,26 +130,99 @@ type goPlayer
       
       if sgnBefore <> sgnAfter then me.spin = 0
     end function
+  
+  function tryCharge() as Boolean
+ 
+    me.tryChargeChronos.raz
 
-    function tryThrust(boost as Float32)
-      dim context as GameContext at me.contextPtr
-      me.thrust += boost / context.frameRate
-      
-      if me.thrust > 10 then me.thrust = 10
-      me.thrustImpulse = true
-      
-      me.alterEnergy(-1)
-    end function
+    me.charge = me.tryChargeChronos.elapsedSinceStart
+    me.alterEnergy(-1/me.tryChargeChronos.framerate)
+    
+    if me.charge >= %MAXCHARGEDURATION Then ' Charge is full
+      if %LOGGERPS then me.logger.write("Full charge reached")
+      me.charge = %MAXCHARGEDURATION
+      me.tryChargeChronos.unlock ' reset charging time (for charge full)
+      return %TRUE ' launching jump
+    Else
+      return %FALSE
+    EndIf
+  
+  end function
+  
+  Function tryThrust() as Boolean
+  
+    Local easeCache as Double
+    
+    if not me.tryThrustChronos.locked then
+      me.tryThrustChronos.raz
+      me.currentCharge = %MAXCHARGE * Pow((me.charge)/(%MAXCHARGEDURATION),(1/%CHARGEPROGRESSIVENESS))
+      me.currentThrustDuration = %THRUSTDURATION * me.charge / %MAXCHARGEDURATION
+      me.tryChargeChronos.unlock
+    Else ' Already in travel
+      me.temps += me.tryThrustChronos.elapsedSinceLastCall
+    end if
 
-    function tryFadeThrust(brake as Float32)
-      dim context as GameContext at me.contextPtr
-      me.thrust -= brake / context.frameRate
-      
-      if me.thrust < 0 then
-        me.thrust = 0
-        me.thrustImpulse = false
-      end if
-    end function
+    if me.temps > me.currentThrustDuration then ' end of travel reached
+      me.thrust = 0
+      me.temps = 0
+      me.currentThrustDuration = 0
+      me.tryThrustChronos.unlock
+      me.thrustImpulse = %FALSE
+    elseif me.currentThrustDuration > 0 then
+      if me.thrust < %MIN_BURST_SPEED Then function = %TRUE ' speed is low enough to not burst bubbles
+      easeCache = (me.easeSimple(me.temps, me.currentCharge , me.currentThrustDuration) + me.crumb)
+      me.crumb = frac(easeCache) ' maybe not useful in the end ?
+      me.thrust = fix(easeCache)
+      me.thrustImpulse = %TRUE
+    endif
+
+  end function
+'[todo] First lvl : The Pilgrim brake to a stop between each jump. (charge - jump - charge - jump ...)
+'       Feat unlock : the Pilgrim can tie each jump to the previous (like swimming : charge - jump and charge while gliding along - jump and charge while gliding along - jump and charge while gliding along)
```

Comment:
@petrSchreiber , f you're ok with that, I would prefer work on this in another feat_branch while merging this one.

Reply:
Sure!